File: src/lib/historyMonthlyAggregation.ts
import { isValidSessionCalendarDate } from "./sessionCalendar";
import { sessionTimeSeconds, type HistorySessionTimeInput } from "./historyStats";

export type MonthlySummary = {
  yearMonth: string;
  label: string;
  daysActive: number;
  daysInMonth: number;
  totalTimeSeconds: number;
};

function parseYearMonth(isoDate: string): { year: number; month: number } {
  const [y, m] = isoDate.split("-").map(Number);
  return { year: y, month: m };
}

export function yearMonthKey(year: number, month: number): string {
  return `${year}-${String(month).padStart(2, "0")}`;
}

function daysInCalendarMonth(year: number, month: number): number {
  return new Date(Date.UTC(year, month, 0)).getUTCDate();
}

function monthLabel(year: number, month: number): string {
  const d = new Date(Date.UTC(year, month - 1, 1, 12));
  return d.toLocaleDateString("en-US", { month: "long", year: "numeric", timeZone: "UTC" });
}

function shortMonthLabel(year: number, month: number): string {
  const d = new Date(Date.UTC(year, month - 1, 1, 12));
  return d.toLocaleDateString("en-US", { month: "short", year: "numeric", timeZone: "UTC" });
}

function groupSessionsByDate(
  sessions: HistorySessionTimeInput[],
): Map<string, HistorySessionTimeInput[]> {
  const map = new Map<string, HistorySessionTimeInput[]>();
  for (const session of sessions) {
    if (!isValidSessionCalendarDate(session.sessionDate)) continue;
    const date = session.sessionDate!;
    (map.get(date) ?? map.set(date, []).get(date)!).push(session);
  }
  return map;
}

function advanceMonth(year: number, month: number, delta: number): { year: number; month: number } {
  let y = year;
  let m = month + delta;
  while (m < 1) {
    m += 12;
    y--;
  }
  while (m > 12) {
    m -= 12;
    y++;
  }
  return { year: y, month: m };
}

/** Aggregate days-active and total sit time for one calendar month. */
export function aggregateMonthSummary(
  sessions: HistorySessionTimeInput[],
  year: number,
  month: number,
): MonthlySummary {
  const ym = yearMonthKey(year, month);
  const byDate = groupSessionsByDate(sessions);
  const dim = daysInCalendarMonth(year, month);
  let daysActive = 0;
  let totalTimeSeconds = 0;

  for (let day = 1; day <= dim; day++) {
    const dd = String(day).padStart(2, "0");
    const isoDate = `${ym}-${dd}`;
    const daySessions = byDate.get(isoDate);
    if (daySessions && daySessions.length > 0) {
      daysActive++;
      for (const s of daySessions) {
        totalTimeSeconds += sessionTimeSeconds(s);
      }
    }
  }

  return {
    yearMonth: ym,
    label: monthLabel(year, month),
    daysActive,
    daysInMonth: dim,
    totalTimeSeconds,
  };
}

/**
 * Trailing 12 calendar months ending with the month containing `todayIso`.
 * Used by the year-in-review grid (#380).
 */
export function buildTrailing12MonthSummaries(
  sessions: HistorySessionTimeInput[],
  todayIso: string,
): MonthlySummary[] {
  const { year: todayYear, month: todayMonth } = parseYearMonth(todayIso);
  const start = advanceMonth(todayYear, todayMonth, -11);

  const summaries: MonthlySummary[] = [];
  let y = start.year;
  let m = start.month;

  for (let i = 0; i < 12; i++) {
    summaries.push(aggregateMonthSummary(sessions, y, m));
    ({ year: y, month: m } = advanceMonth(y, m, 1));
  }

  return summaries;
}

/** Short month label for compact grid cells (e.g. "Jul 2026"). */
export function formatShortMonthLabel(yearMonth: string): string {
  const { year, month } = parseYearMonth(`${yearMonth}-01`);
  return shortMonthLabel(year, month);
}

/**
 * One compact row per calendar month strictly before the month of `todayIso`.
 * Spans from the earliest session month through the month before current.
 */
export function buildPriorMonthSummaries(
  sessions: HistorySessionTimeInput[],
  todayIso: string,
): MonthlySummary[] {
  const { year: todayYear, month: todayMonth } = parseYearMonth(todayIso);
  const currentYm = yearMonthKey(todayYear, todayMonth);

  const validDates = sessions
    .map(s => s.sessionDate)
    .filter((d): d is string => isValidSessionCalendarDate(d));
  if (validDates.length === 0) return [];

  const earliest = validDates.reduce((a, b) => (a < b ? a : b));
  const { year: startYear, month: startMonth } = parseYearMonth(earliest);

  const summaries: MonthlySummary[] = [];
  let y = startYear;
  let m = startMonth;

  while (true) {
    const ym = yearMonthKey(y, m);
    if (ym >= currentYm) break;

    summaries.push(aggregateMonthSummary(sessions, y, m));
    ({ year: y, month: m } = advanceMonth(y, m, 1));
  }

  return summaries;
}

File: ios/StillPointShared/Sources/StillPointShared/HistoryMonthlyAggregation.swift
import Foundation

public struct MonthlySummary: Sendable, Equatable {
    public let yearMonth: String
    public let label: String
    public let daysActive: Int
    public let daysInMonth: Int
    public let totalTimeSeconds: Int

    public init(
        yearMonth: String,
        label: String,
        daysActive: Int,
        daysInMonth: Int,
        totalTimeSeconds: Int
    ) {
        self.yearMonth = yearMonth
        self.label = label
        self.daysActive = daysActive
        self.daysInMonth = daysInMonth
        self.totalTimeSeconds = totalTimeSeconds
    }
}

public enum HistoryMonthlyAggregation {
    private static let monthLabelFormatter: DateFormatter = {
        let df = DateFormatter()
        df.locale = Locale(identifier: "en_US_POSIX")
        df.timeZone = TimeZone(secondsFromGMT: 0)
        df.dateFormat = "MMMM yyyy"
        return df
    }()

    private static let shortMonthLabelFormatter: DateFormatter = {
        let df = DateFormatter()
        df.locale = Locale(identifier: "en_US_POSIX")
        df.timeZone = TimeZone(secondsFromGMT: 0)
        df.dateFormat = "MMM yyyy"
        return df
    }()

    public static func yearMonthKey(year: Int, month: Int) -> String {
        String(format: "%04d-%02d", year, month)
    }

    private static func parseYearMonth(_ isoDate: String) -> (year: Int, month: Int) {
        let parts = isoDate.split(separator: "-").compactMap { Int($0) }
        guard parts.count >= 2 else { return (0, 0) }
        return (parts[0], parts[1])
    }

    private static func daysInCalendarMonth(year: Int, month: Int) -> Int {
        var components = DateComponents()
        components.year = year
        components.month = month
        components.day = 1
        components.calendar = utcCalendar
        components.timeZone = TimeZone(secondsFromGMT: 0)
        guard let date = utcCalendar.date(from: components),
              let range = utcCalendar.range(of: .day, in: .month, for: date) else { return 30 }
        return range.count
    }

    private static let utcCalendar: Calendar = {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(secondsFromGMT: 0)!
        return calendar
    }()

    private static func monthLabel(year: Int, month: Int) -> String {
        var components = DateComponents()
        components.year = year
        components.month = month
        components.day = 1
        components.hour = 12
        components.calendar = utcCalendar
        components.timeZone = TimeZone(secondsFromGMT: 0)
        guard let date = utcCalendar.date(from: components) else { return yearMonthKey(year: year, month: month) }
        return monthLabelFormatter.string(from: date)
    }

    private static func shortMonthLabel(year: Int, month: Int) -> String {
        var components = DateComponents()
        components.year = year
        components.month = month
        components.day = 1
        components.hour = 12
        components.calendar = utcCalendar
        components.timeZone = TimeZone(secondsFromGMT: 0)
        guard let date = utcCalendar.date(from: components) else { return yearMonthKey(year: year, month: month) }
        return shortMonthLabelFormatter.string(from: date)
    }

    private static func groupAnySessionsByDate(
        _ sessions: [some HistorySessionTimeInput]
    ) -> [String: [any HistorySessionTimeInput]] {
        var map: [String: [any HistorySessionTimeInput]] = [:]
        for session in sessions {
            guard SessionCalendar.isValidSessionCalendarDate(session.sessionDate) else { continue }
            map[session.sessionDate, default: []].append(session)
        }
        return map
    }

    private static func aggregateMonthSummaryFromGrouped(
        byDate: [String: [any HistorySessionTimeInput]],
        year: Int,
        month: Int
    ) -> MonthlySummary {
        let ym = yearMonthKey(year: year, month: month)
        let dim = daysInCalendarMonth(year: year, month: month)
        var daysActive = 0
        var totalTimeSeconds = 0

        for day in 1...dim {
            let dd = String(format: "%02d", day)
            let isoDate = "\(ym)-\(dd)"
            guard let daySessions = byDate[isoDate], !daySessions.isEmpty else { continue }
            daysActive += 1
            for session in daySessions {
                totalTimeSeconds += HistorySessionTime.sessionTimeSeconds(session)
            }
        }

        return MonthlySummary(
            yearMonth: ym,
            label: monthLabel(year: year, month: month),
            daysActive: daysActive,
            daysInMonth: dim,
            totalTimeSeconds: totalTimeSeconds
        )
    }

    public static func aggregateMonthSummary(
        sessions: [some HistorySessionTimeInput],
        year: Int,
        month: Int
    ) -> MonthlySummary {
        aggregateMonthSummaryFromGrouped(
            byDate: groupAnySessionsByDate(sessions),
            year: year,
            month: month
        )
    }

    private static func advanceMonth(year: Int, month: Int, delta: Int) -> (year: Int, month: Int) {
        var y = year
        var m = month + delta
        while m < 1 {
            m += 12
            y -= 1
        }
        while m > 12 {
            m -= 12
            y += 1
        }
        return (y, m)
    }

    /// Trailing 12 calendar months ending with the month containing `todayIso`.
    public static func buildTrailing12MonthSummaries(
        sessions: [some HistorySessionTimeInput],
        todayIso: String
    ) -> [MonthlySummary] {
        let (todayYear, todayMonth) = parseYearMonth(todayIso)
        let start = advanceMonth(year: todayYear, month: todayMonth, delta: -11)
        let byDate = groupAnySessionsByDate(sessions)

        var summaries: [MonthlySummary] = []
        var y = start.year
        var m = start.month

        for _ in 0..<12 {
            summaries.append(aggregateMonthSummaryFromGrouped(byDate: byDate, year: y, month: m))
            (y, m) = advanceMonth(year: y, month: m, delta: 1)
        }

        return summaries
    }

    /// Short month label for compact grid cells (e.g. "Jul 2026").
    public static func formatShortMonthLabel(_ yearMonth: String) -> String {
        let (year, month) = parseYearMonth("\(yearMonth)-01")
        return shortMonthLabel(year: year, month: month)
    }

    /// One compact row per calendar month strictly before the month of `todayIso`.
    public static func buildPriorMonthSummaries(
        sessions: [some HistorySessionTimeInput],
        todayIso: String
    ) -> [MonthlySummary] {
        let (todayYear, todayMonth) = parseYearMonth(todayIso)
        let currentYm = yearMonthKey(year: todayYear, month: todayMonth)

        let validDates = sessions
            .map(\.sessionDate)
            .filter(SessionCalendar.isValidSessionCalendarDate)
        guard let earliest = validDates.min() else { return [] }

        let (startYear, startMonth) = parseYearMonth(earliest)
        let byDate = groupAnySessionsByDate(sessions)

        var summaries: [MonthlySummary] = []
        var y = startYear
        var m = startMonth

        while true {
            let ym = yearMonthKey(year: y, month: m)
            if ym >= currentYm { break }
            summaries.append(aggregateMonthSummaryFromGrouped(byDate: byDate, year: y, month: m))
            (y, m) = advanceMonth(year: y, month: m, delta: 1)
        }

        return summaries
    }
}
